File: app/src/java/java/ru/esm/tspiot/dkkt/service/TrustedEncryptionResult.java
package ru.esm.tspiot.dkkt.service;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;

import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.Objects;

public class TrustedEncryptionResult implements Parcelable {
    private final byte[] mControl;
    private final byte[] mData;

    public TrustedEncryptionResult(byte[] mControl, byte[] mData) {
        this.mControl = mControl;
        this.mData = mData;
    }

    public byte[] getControl() {
        return mControl;
    }

    public byte[] getData() {
        return mData;
    }

    @Override
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) return false;
        TrustedEncryptionResult that = (TrustedEncryptionResult) o;
        return Objects.deepEquals(mControl, that.mControl) && Objects.deepEquals(mData, that.mData);
    }

    @Override
    public int hashCode() {
        return Objects.hash(Arrays.hashCode(mControl), Arrays.hashCode(mData));
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(@NonNull @NotNull Parcel dest, int flags) {
        dest.writeByteArray(mControl);
        dest.writeByteArray(mData);
    }

    public static final Creator<TrustedEncryptionResult> CREATOR = new Creator<>() {
        @Override
        public TrustedEncryptionResult createFromParcel(Parcel in) {
            return new TrustedEncryptionResult(in.createByteArray(), in.createByteArray());
        }

        @Override
        public TrustedEncryptionResult[] newArray(int size) {
            return new TrustedEncryptionResult[size];
        }
    };
}

File: app/src/kotlin/kotlin/ru/esm/tspiot/dkkt/service/TrustedEncryptionResult.kt
package ru.esm.tspiot.dkkt.service

import android.os.Parcel
import android.os.Parcelable

data class TrustedEncryptionResult(
    val encryptedData: ByteArray,
    val controlCode: ByteArray,
    val statusCode: Int,
    val statusMessage: String
) : Parcelable {
    constructor(parcel: Parcel) : this(
        parcel.createByteArray() ?: byteArrayOf(),
        parcel.createByteArray() ?: byteArrayOf(),
        parcel.readInt(),
        parcel.readString() ?: ""
    )

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeByteArray(encryptedData)
        parcel.writeByteArray(controlCode)
        parcel.writeInt(statusCode)
        parcel.writeString(statusMessage)
    }

    override fun describeContents(): Int = 0

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as TrustedEncryptionResult

        if (!encryptedData.contentEquals(other.encryptedData)) return false
        if (!controlCode.contentEquals(other.controlCode)) return false
        if (statusCode != other.statusCode) return false
        if (statusMessage != other.statusMessage) return false

        return true
    }

    override fun hashCode(): Int {
        var result = encryptedData.contentHashCode()
        result = 31 * result + controlCode.contentHashCode()
        result = 31 * result + statusCode
        result = 31 * result + statusMessage.hashCode()
        return result
    }

    companion object CREATOR : Parcelable.Creator<TrustedEncryptionResult> {
        override fun createFromParcel(parcel: Parcel): TrustedEncryptionResult =
            TrustedEncryptionResult(parcel)

        override fun newArray(size: Int): Array<TrustedEncryptionResult?> =
            arrayOfNulls(size)
    }
}